src/test/store/receive-profile.test.js

```diff
@@ -522,17 +537,11 @@ describe('actions/receive-profile', function() {
     }: any): Response);
 
     beforeEach(function() {
-      // The stub makes it easy to return different values for different
-      // arguments. Here we define the default return value because there is no
-      // argument specified.
       window.fetch = jest.fn().mockResolvedValue(fetch403Response);
-
-      sinon.stub(window, 'setTimeout').yieldsAsync(); // will call its argument asynchronously
```

Comment:
So, the goal of this line is to call the argument of setTimeout asynchronously right away (instead of waiting for the timeout). I believe we could do something like this with jest (without using fake timers):
```
jest.spyOn(window, 'setTimeout').mockImplementation(callback => process.nextTick(callback))
```
or
```
jest.spyOn(window, 'setTimeout').mockImplementation(callback => setImmediate(callback))
```

Reply:
I was struggling with these too. At first, none of them would work and I'd get the message
`Timeout - Async callback was not invoked within the 5000ms timeout specified by jest.setTimeout.` It's the same error I had until I found the double resolve solution.

Did some experimenting, but no progress until I by chance added back `jest.useFakeTimers()`. Then it suddenly worked(!).

So picked your first suggestion, which is way clearer than my first solution, but I also call `jest.useFakeTimers()` first to make it work. Quite unsettling to not know why it's needed though.

Reply:
So, this is very interesting. I tried removing `useFakeTimers` to debug. I found like you that it doesn't work and I get a timeout in that case. _But_ I also found that if I execute only one suite (using `describe.only`) then it runs properly!
So there's a weird interaction between suites... I'm still debugging.

Reply:
I found the problem, here is a simple patch:
```diff
diff --git a/src/test/components/ProfileCallTreeView.test.js b/src/test/components/ProfileCallTreeView.test.js
index 77e436a4..76b2716a 100644
--- a/src/test/components/ProfileCallTreeView.test.js
+++ b/src/test/components/ProfileCallTreeView.test.js
@@ -482,16 +482,22 @@ describe('calltree/ProfileCallTreeView TransformNavigator', () => {
   });
 });
 
 describe('ProfileCallTreeView/end-to-end', () => {
   // This next test explicitly does not have an assertion, disable the eslint rule
   // requiring one.
   // eslint-disable-next-line jest/expect-expect
   it('can display a gecko profile without crashing', () => {
+    // We use the fake indexeddb in this test because the symbolication kicks
+    // in. The fake indexeddb relies on setImmediate to run events. That's why
+    // we use useFakeTimers so that pending event handlers are canceled when the
+    // test runs and we don't get spurious messages.
+    jest.useFakeTimers();
+
     const geckoProfile = createGeckoProfile();
     const processedProfile = processProfile(geckoProfile);
     const store = storeWithProfile(processedProfile);
     render(
       <Provider store={store}>
         <ProfileCallTreeView />
       </Provider>
     );
diff --git a/src/test/setup.js b/src/test/setup.js
index 97ed83d9..984a75bc 100644
--- a/src/test/setup.js
+++ b/src/test/setup.js
@@ -20,16 +20,17 @@ afterEach(cleanup);
 afterEach(() => {
   // The configuration to restore and reset all of the mocks in tests does not seem
   // to be working correctly. Manually trigger this call here to ensure we
   // don't get intermittents from one test's mocks affecting another test's mocks.
   //
   // See https://github.com/facebook/jest/issues/7654
   jest.resetAllMocks();
   jest.clearAllTimers();
+  jest.useRealTimers();
 });
 
 expect.extend({
   toHaveClass(received, className) {
     if (!(received instanceof Element)) {
       throw new Error(
         `expected value ${received} to be an instance of Element.`
       );
diff --git a/src/test/store/receive-profile.test.js b/src/test/store/receive-profile.test.js
index 1fd183a6..aa9b701e 100644
--- a/src/test/store/receive-profile.test.js
+++ b/src/test/store/receive-profile.test.js
@@ -517,17 +517,17 @@ describe('actions/receive-profile', function() {
         get: () => 'application/json',
       },
       json: () =>
         Promise.resolve(JSON.parse(serializeProfile(_getSimpleProfile()))),
     }: any): Response);
 
     beforeEach(function() {
       window.fetch = jest.fn().mockResolvedValue(fetch403Response);
-      jest.useFakeTimers();
+
       // Call the argument of setTimeout asynchronously right away
       // (instead of waiting for the timeout).
       jest
         .spyOn(window, 'setTimeout')
         .mockImplementation(callback => process.nextTick(callback));
     });
 
     afterEach(function() {
@@ -673,17 +673,17 @@ describe('actions/receive-profile', function() {
         get: () => 'application/json',
       },
       json: () =>
         Promise.resolve(JSON.parse(serializeProfile(_getSimpleProfile()))),
     }: any): Response);
 
     beforeEach(function() {
       window.fetch = jest.fn().mockResolvedValue(fetch403Response);
-      jest.useFakeTimers();
+
       // Call the argument of setTimeout asynchronously right away
       // (instead of waiting for the timeout).
       jest
         .spyOn(window, 'setTimeout')
         .mockImplementation(callback => process.nextTick(callback));
     });
 
     afterEach(function() {
```

The problem was that jest doesn't automatically reset the "fake timers" between 2 tests... I believe it does it between 2 files because it uses a complete new environment.

I had to add useFakeTimers in another test that actually needs it. I added a comment to explain why.

Reply:
Wow, great finding! I'll check it out and apply your patch. Thanks!